sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala

```diff
@@ -170,10 +232,27 @@ class TungstenAggregationIterator(
     val bufferRowSize: Int = bufferSchema.length
 
     val genericMutableBuffer = new GenericMutableRow(bufferRowSize)
-    val unsafeProjection =
-      UnsafeProjection.create(bufferSchema.map(_.dataType))
-    val buffer = unsafeProjection.apply(genericMutableBuffer)
-    initialProjection.target(buffer)(EmptyRow)
+    // TODO(josh): figure out whether we have to use
+    val useUnsafeBuffer = bufferSchema.map(_.dataType).forall(UnsafeRow.isMutable)
+
+    val buffer =  /* if (useUnsafeBuffer) */ {
+      val unsafeProjection =
+        UnsafeProjection.create(bufferSchema.map(_.dataType))
+      unsafeProjection.apply(genericMutableBuffer)
+//    } else {
+//      genericMutableBuffer
+    }
+    expressionAggInitialProjection.target(buffer)(EmptyRow)
+    // TODO(josh): this can be done more cleanly
```

Comment:
I think that I thought that this block of code was a bit messy while I was working on it. Will clean up while addressing any code review comments.